client/files.go

```diff
@@ -0,0 +1,203 @@
+	// instead have their information returned as if they were regular files.
+	Itself bool
+}
+
+type FileInfo struct {
+	name    string
+	size    int64
+	mode    os.FileMode
+	modTime time.Time
+	path    string
+	userID  *int
+	groupID *int
+	user    string
+	group   string
+}
+
+// Name returns the base name of the file.
+func (fi *FileInfo) Name() string {
+	return fi.name
+}
+
+// Size returns the length in bytes for regular files. For others, its
+// behavior is system-dependent.
+func (fi *FileInfo) Size() int64 {
+	return fi.size
+}
+
+// Mode returns the file mode and permission bits.
+func (fi *FileInfo) Mode() os.FileMode {
+	return fi.mode
+}
+
+// ModTime returns the file modification time.
+func (fi *FileInfo) ModTime() time.Time {
+	return fi.modTime
+}
+
+// IsDir is an abbreviation for Mode().IsDir().
+func (fi *FileInfo) IsDir() bool {
+	return fi.mode.IsDir()
+}
+
+// Sys returns the underlying data source (always nil for client.FileInfo).
+func (fi *FileInfo) Sys() interface{} {
+	return nil
+}
+
+// Path is the full absolute path of the file.
+func (fi *FileInfo) Path() string {
+	return fi.path
+}
+
+// UserID is the ID of the owner user (can be nil).
+func (fi *FileInfo) UserID() *int {
+	return fi.userID
+}
+
+// GroupID is the ID of the owner group (can be nil).
+func (fi *FileInfo) GroupID() *int {
+	return fi.groupID
+}
+
+// User is the string representing the owner user name.
+func (fi *FileInfo) User() string {
+	return fi.user
+}
+
+// Group is the string representing the owner user group.
+func (fi *FileInfo) Group() string {
+	return fi.group
+}
+
+// ListFiles obtains the contents of a directory or glob, or information about a file.
+func (client *Client) ListFiles(opts *ListFilesOptions) ([]*FileInfo, error) {
+	q := make(url.Values)
+	q.Set("action", "list")
+	q.Set("path", opts.Path)
+	if opts.Pattern != "" {
+		q.Set("pattern", opts.Pattern)
+	}
+	if opts.Itself {
+		q.Set("itself", "true")
+	}
+
+	var results []fileInfoResult
+	_, err := client.doSync("GET", "/v1/files", q, nil, nil, &results)
+	if err != nil {
+		return nil, err
+	}
+
+	infos := make([]*FileInfo, len(results))
+	for i, result := range results {
+		infos[i], err = resultToFileInfo(result)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return infos, nil
+}
+
+type fileInfoResult struct {
+	Path         string `json:"path"`
+	Name         string `json:"name"`
+	Type         string `json:"type"`
+	Size         *int64 `json:"size,omitempty"`
+	Permissions  string `json:"permissions"`
+	LastModified string `json:"last-modified"`
+	UserID       *int   `json:"user-id"`
+	User         string `json:"user"`
+	GroupID      *int   `json:"group-id"`
+	Group        string `json:"group"`
+}
+
+func calculateFileMode(fileType string, permissions string) (mode os.FileMode, err error) {
+	p, err := strconv.ParseUint(permissions, 8, 32)
+	if err != nil {
+		return 0, err
+	}
+
+	mode = os.FileMode(p) & os.ModePerm
```

Comment:
We are masking out all bits here except for 0777. I just want to make sure its intentional that we do not want to show setuid, setgid or sticky.

Reply:
I did not take that into account! We should fix this along [the 4-digit octal issue you mentioned in the other PR](https://github.com/canonical/pebble/pull/144#discussion_r959414764).